File: machine/ts/machine/devices/audio/badp_decoder.ts
import { readLE32 } from '../../../common/endian';

export type ApuBadpSeekTable = {
	bytes: Uint8Array;
	byteOffset: number;
	entryCount: number;
};

export type ApuBadpDecoderState = {
	predictors: Int32Array;
	stepIndices: Int32Array;
	nextFrame: number;
	blockEnd: number;
	blockFrames: number;
	blockFrameIndex: number;
	payloadOffset: number;
	nibbleCursor: number;
	decodedFrame: number;
	decodedLeft: number;
	decodedRight: number;
	previousDecodedFrame: number;
	previousDecodedLeft: number;
	previousDecodedRight: number;
};

export function loadApuBadpSeekTable(out: ApuBadpSeekTable, bytes: Uint8Array, byteOffset: number): void {
	out.bytes = bytes;
	out.byteOffset = byteOffset + readLE32(bytes, byteOffset + 32);
	out.entryCount = readLE32(bytes, byteOffset + 28);
}

File: machine/cpp/machine/devices/audio/badp_decoder.cpp
/*
 * badp_decoder.cpp - APU BADP block decoder and seek table parsing.
 */

#include "machine/devices/audio/badp_decoder.h"

#include "common/endian.h"

namespace bmsx {

void loadApuBadpSeekTable(ApuBadpSeekTable& out, const u8* bytes, size_t byteOffset) {
	out.bytes = bytes;
	out.byteOffset = byteOffset + readLE32(bytes + byteOffset + 32u);
	out.entryCount = readLE32(bytes + byteOffset + 28u);
}

} // namespace bmsx
